content/blog/blog-2020-07-10-webhook-development.md

```diff
@@ -0,0 +1,153 @@
+---
+date: 2020-07-10
+---
+
+# Developing and Testing Kubernetes Webhooks
+
+Kubernetes development, testing and debugging can be challenging by itself, but fewer things are more challenging than debugging webhooks or technologies like them.  Specifically components which require a callback AND that callback requires a valid certificate to succeeded.  See requests for [Come up with webhook developer workflow to test it locally](https://github.com/kubernetes-sigs/kubebuilder/issues/400), I thought it would be valuable to the community to post a solution.  
+
+Working on KUDO we have this concern as well and thought it would be worth showing a technique we use to debug locally.  While this post uses KUDO as an example, it could be any Kubernetes webhook.  It is worth pointing out that KUDO uses [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) under the hood which is the component that is actually listening on port 443.  The current optimized solution requires [ngrok](https://ngrok.com/) which requires an account registration for working with https port 443.
+
+<!-- more -->
+
+## Local Development Env and Assumptions
+
+We commonly work with [kind](https://kind.sigs.k8s.io/) for local development and will assume that through the article. For the rest of the article we will assume that `kind create cluster` was used to create a new local cluster and that you have access to that cluster with `kubectl`
+
+For KUDO, the manager is normally deployed into a cluster as a pod with a container from `kudobuilder/controller:v0.15.0`.  This manager uses controller-runtime and registers a kudo controller and a `MutatingWebhookConfiguration` webhook.  The kubernetes configuration for this can be viewed in kudo by dumping the prerequisite manifests generated by the CLI's `init` command.  `go run cmd/kubectl-kudo/main.go init --unsafe-self-signed-webhook-ca --version dev --dry-run -o yaml` will show all manifests including `kind: MutatingWebhookConfiguration`.
```

Comment:
```suggestion
For KUDO, the manager is normally deployed into a cluster as a pod with a container from `kudobuilder/controller:v0.15.0`.  This manager uses controller-runtime and registers a kudo controller and a `MutatingWebhookConfiguration` webhook.  The kubernetes configuration for this can be viewed in KUDO by dumping the prerequisite manifests generated by the CLI's `init` command.  `go run cmd/kubectl-kudo/main.go init --unsafe-self-signed-webhook-ca --version dev --dry-run -o yaml` will show all manifests including `kind: MutatingWebhookConfiguration`.
```